### PR TITLE
Email system tff 175

### PR DESCRIPTION
**Pull request by Cafferoq**
### Summary

System to send emails to people.

### Checklist

- [x] Have you added unit and functional tests as appropriate?
- [x] Did you update/add documentation for new methods or changed functionality?
- [x] Have you merged the target branch down into this one?

### How to Review

1. Pull
2. Make sure tests work.
3. Make sure code makes sense. 

### Links

- [TFF-175](https://msoese.atlassian.net/browse/TFF-175)

### CODE REVIEW

File: src/lib/clock.js [hunk 1]
```diff
@@ -15,9 +16,10 @@ function getInitialTimeoutTime(){
 }
 
 setTimeout(() => {
-    //TODO: connect email system here
     //email Students
-    //setInterval(/*Email Students*/, 86400000);
+    sendDueItemNotifications();
+    //set-up timeout to send notifications every 24 hours.
+    setInterval(sendDueItemNotifications, 86400000);
```
Comment (thofmann) [inline src/lib/clock.js]: Nit: I would rather write this as `24*60*60*1000` for clarity.

File: src/lib/email.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import StudentStore from '../store/student-store';
+
+const mailer = sendmail({
+    silent: true
+});
+
+export function sendMail(emails){
+    return new Promise((res, rej) => {
+        mailer({
+            from: 'testing@msoe.edu',
```
Comment (thofmann) [inline src/lib/email.js]: NACK: Should this be something like `techsupport@msoe.edu`? We should also include the `replyTo` option with `'no-reply@msoe.edu'` or Jim's email address. https://github.com/guileen/node-sendmail#mail-options

File: src/lib/email.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+            to: emails.join(','),
+            subject: 'ITEMS DUE SOON',
```
Comment (thofmann) [inline src/lib/email.js]: Nit: This subject line looks spammy. Maybe `'MSOE Tech Support Items Due Soon'` or just `'Items Due Soon'`?

File: src/lib/email.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+            html: 'This is a test.'
```
Comment (thofmann) [inline src/lib/email.js]: NACK: This is no longer a test. Should we include a list of the item addresses / model names?

File: src/lib/email.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        }, err => {
+            if (err) return rej(err);
+            else res();
+        });
+    });
+}
+
+export function sendDueItemNotifications(){
+    let emails = StudentStore.getStudentsWithDueItems().map(student => student.email);
+    return this.sendMail(emails);
```
Comment (thofmann) [inline src/lib/email.js]: Nit: Do we want to send one email to all of the students? Seems sub-optimal for privacy for students to be able to see the full list of other students with due items, but not too bad.

Comment (thofmann) [inline src/lib/email.js]: NACK: Should we also include the items that are overdue?

File: src/store/student-store.js [hunk 6]
```diff
@@ -43,6 +44,15 @@ class StudentStore extends Store {
         return studentsByActionId[actionId];
     }
 
+    getStudentsWithDueItems(){
+        return students.filter(student => {
+            return student.items.some(item => {
+                return isBeforeNow(item.timestamp - 1800);
+                //subtracting 1800 to see if it's due in the next half hour
```
Comment (thofmann) [inline src/store/student-store.js]: Very minor nit: Comment before code you're describing (or on the same line).

File: test/unit/lib/email.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import StudentStore from '../../../.dist/store/student-store';
+import * as email from '../../../.dist/lib/email';
+import sinon from 'sinon';
+
+describe("Email system", () => {
+    it("Sends an email to the correct students", () => {
+        let getStudentsWithDueItems = sinon.stub(StudentStore, "getStudentsWithDueItems");
+
+        getStudentsWithDueItems.returns([
+            {email: 'trevinhofmann@gmail.com'}
```
Comment (thofmann) [inline test/unit/lib/email.js]: :+1:

File: src/lib/email.js (hunk 2 above)
Comment (Cafferoq) [inline src/lib/email.js]: I was just going to leave that as is and go back to set the details after we talk to Jim, that goes for most of the comments in the review, but I'll go change it now if you feel it's important to do right away.

File: src/lib/email.js (hunk 4 above)
Comment (Cafferoq) [inline src/lib/email.js]: If we want that I feel we should make it another task, as that would take substantially more work.

File: src/lib/email.js (hunk 5 above)
Comment (Cafferoq) [inline src/lib/email.js]: I didn't really think about that. We should probably split that up. Also, see previous comment about adding another task if we want overdue items to come with them.

**Comment by thofmann** [comment]
utACK

File: src/lib/email.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import sendmail from 'sendmail';
+import StudentStore from '../store/student-store';
+
+const mailer = sendmail({
+    silent: true
+});
+
+export function sendMail(emails){
+    var promises = [];
```
Comment (gearedmonkey) [inline src/lib/email.js]: change to let

File: test/unit/lib/email.js (hunk 7 above)
Comment (gearedmonkey) [inline test/unit/lib/email.js]: Changing the amount of emails that this method returns makes the 'should get students with soon to be due items' test fail. Might have something to do with you not restoring the spy?

**Pull request closed without merging**